tests/python/test_communication.py

```diff
@@ -15,32 +15,87 @@
 
 @pytest.mark.mpi
 def test_allgather(mpi_test):
-    num_devices = mpi_test.size
-    rank = mpi_test.rank
-
-    unsharded = torch.randn(num_devices * 4)
-    sharded = mpi_test.shard_tensor(unsharded, 0)
+    d = mpi_test.size
 
     class Model(FusionDefinition):
         def definition(self):
             self.inp = self.define_tensor(
-                (num_devices * 4,), contiguity=True, dtype=DataType.Float
+                (d * 4,), contiguity=True, dtype=DataType.Float
             )
             self.out = self.ops.set(self.inp)
             self.add_output(self.out)
 
         def multidevice_schedule(self):
-            mesh = self.sched._create_device_mesh(range(num_devices))
+            mesh = self.sched._create_device_mesh(range(d))
             self.sched._set_device_mesh(self.inp, mesh)
             self.sched._set_device_mesh(self.out, mesh)
 
-            self.sched.split(self.inp, 0, num_devices, False)
+            self.sched.split(self.inp, 0, d, False)
             self.sched.parallelize(self.inp, 0, nvfuser.ParallelType.mesh_x)
             self.sched.set_allocation_as_loop(self.inp)
 
-            self.sched.split(self.out, 0, num_devices, False)
+            self.sched.split(self.out, 0, d, False)
             self.sched.set_allocation_as_loop(self.out)
 
+    unsharded = torch.randn(d * 4)
+    sharded = mpi_test.shard_tensor(unsharded, 0)
+
     fd = Model()
     outputs = fd.execute([sharded])
     torch.testing.assert_close(outputs[0].cpu(), unsharded)
+
+
+@pytest.mark.mpi
+def test_allreduce(mpi_test):
```

Comment:
This allreduce test is merely for DID logical split. I don't think allreduce can support DID loop split because `sum`'s reduction axes can only be logical. But I'd be happy to know otherwise.

Reply:
> This allreduce test is merely for DID logical split.

Just to be clear, you meant DID parallelization of logical domains, right? I'm not sure what you meant by DID logical split otherwise.

Reply:
Assuming I understand what you meant correctly, I think this is where `TensorView::rFactor` could be used. That's what we use for intra-device hierarchical reductions. For example, I'd think that for multi-GPU reductions, we would have something like:

(I'm mixing the C++ and Python APIs)
```
self.out->split(0, num_devices, /*inner=*/false);
auto intermediate_result = self.out->rFactor({1});
intermediate_result->axis(0)->parallelize(DIDx);
self.out->axis(0)->parallelize(DIDx);
```

Here, `intermediate_result` would be the partial result of per-device reduction, which would be then reduced between all the devices and saved to `self.out`.

Reply:
I agree it's something like rfactor, and did look into how TensorView::rfactor works in https://github.com/NVIDIA/Fuser/blob/ecabd46927c9d757c5623e04d35a716e15d3ede3/tests/cpp/test_tutorial.cpp#L344. However, I failed to see how it applies here. 

If we want to loop (but not logical) split an allreduce, the input would be a logical shape like `[D*2,3]` and the output would be of logical shape like `[2,3]`. Regardless of scheduling, what ops in fusion IR could do that? (Not a sum because that reduces an entire dimension to 1).

Reply:
Let's talk offline. It seems we are not using the same vocabulary (e.g., I don't understand what "loop split" and "logical split" mean).

Reply:
https://github.com/NVIDIA/Fuser/pull/3543 is my failed attempt. It triggered an assertion at https://github.com/NVIDIA/Fuser/blob/9346c8ff349a6f09031ecd4deacb424273cf53dd/csrc/scheduler/vectorize_helper.cpp#L1063. Is it because code there has been assuming that reductions are innermost?

Reply:
Anyhow, this isn't a blocker. As we discussed yesterday, we'll probably stick with logical split for reductions in Allreduce and ReduceScatter due to MatmulOp's implementation.